### ISSUE TITLE
User Cannot Reset Password Unless Email Confirmed

### ISSUE THREAD

**Issue by jeffolmstead** [OrchardCMS/OrchardCore#3426]
Perhaps there is a security reason, but I feel it is misleading to not allow a user to reset their password unless they have confirmed their email address. Right now I do NOT have the setting checked that "Users must verify their email address". 

With that being the case, it seems the code in OrchardCore.Users.Controllers.ResetPasswordController --> ForgotPassword shouldn't check to see if the user has already confirmed their email. This is the current code:
```
if (user == null || !(await _userManager.IsEmailConfirmedAsync(user)))
{
     // returns to confirmation page anyway: we don't want to let scrapers know if a username or an email exist
     return RedirectToLocal(Url.Action("ForgotPasswordConfirmation", "ResetPassword"));
}
```
Instead, it seems this check should only be done if the setting is enabled that the users must verify their email address. This is the code change I would propose:
```
var user = await _userService.GetForgotPasswordUserAsync(model.UserIdentifier) as User;
if (user == null || (
        await _siteService.GetSiteSettingsAsync()).As<RegistrationSettings>().UsersMustValidateEmail
        && !await _userManager.IsEmailConfirmedAsync(user)
    )
{
    // returns to confirmation page anyway: we don't want to let scrapers know if a username or an email exist
    return RedirectToLocal(Url.Action("ForgotPasswordConfirmation", "ResetPassword"));
}
```
Look forward to others thoughts.

**Comment by sebastienros** [comment]
Is it fixed by the PR?

**Comment by jeffolmstead** [comment]
@sebastienros no, the prior pull request did not include a fix for this issue. I can quickly implement the pull request utilizing the code above to resolve this issue, was more concerned if anyone could see a security issue for making a change like this. 

I have thought about it more and one of two things should be done. 

1. Either use the above "fix" allowing users to reset their password 
2. OR do not allow users to enable the password reset setting if the Users Must Validate Email Address is not checked (as the password reset won't work).

That is the conclusion I came to.

**Comment by sebastienros** [comment]
The fix looks good

**Comment by jeffolmstead** [comment]
Thinking about this a bit more, I think there might be a problem. The proposed code change is in the OrchardCore.Users.ResetPassword feature while the actual RegistrationSettings being checked on are only editable (created??) via the OrchardCore.Users.Registration feature. Looking at dependencies, someone could have OrchardCore.Users.ResetPassword feature enabled without having enabled OrchardCore.Users.Registration feature. This would seem to cause an issue as "RegistrationSettings" might not even exist. Feedback welcome but i believe I will need to handle null instances like:
```
var user = await _userService.GetForgotPasswordUserAsync(model.UserIdentifier) as User;
if (user == null)
{
    var registrationSettings = (await _siteService.GetSiteSettingsAsync()).As<RegistrationSettings>();
    if (registrationSettings != null
        && registrationSettings.UsersMustValidateEmail
        && !await _userManager.IsEmailConfirmedAsync(user))
    {
        // returns to confirmation page anyway: we don't want to let scrapers know if a username or an email exist
        return RedirectToLocal(Url.Action("ForgotPasswordConfirmation", "ResetPassword"));
    }
}
```
I will have to hand test this in a fresh instance to see what happens when OrchardCore.Users.Registration feature has never been enabled.

**Comment by jeffolmstead** [comment]
@sebastienros I used the fix you said 'looked good' as RegistrationSettings do not come back as null. Also found issue with email not coming in as HTML from template so cleaned that up. Finally, expanded the comment when user requires that an email address be confirmed to add clarity. Let me know if you have questions on the pull request.

**Comment by hishamco** [comment]
@jeffolmstead I think this should be closed while it's covered by your PR #3437

**Comment by jeffolmstead** [comment]
Agreed @hishamco this one is now closed.
